src/streaming/controllers/ExtUrlQueryInfoController.js

```diff
@@ -0,0 +1,195 @@
+ *
+ * Copyright (c) 2013, Dash Industry Forum.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *  * Redistributions of source code must retain the above copyright notice, this
+ *  list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *  this list of conditions and the following disclaimer in the documentation and/or
+ *  other materials provided with the distribution.
+ *  * Neither the name of Dash Industry Forum nor the names of its
+ *  contributors may be used to endorse or promote products derived from this software
+ *  without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS AS IS AND ANY
+ *  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ *  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ *  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ *  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import FactoryMaker from '../../core/FactoryMaker.js';
+import Constants from '../constants/Constants.js';
+
+function ExtUrlQueryInfoController() {
+    let instance,
+        mpd;
+
+
+    function generateQueryParams(resultObject, manifestObejc, propertyString, mpdUrlQuery, defaultInitialString) {
+        const property = manifestObejc[propertyString]?.find((property) => property.schemeIdUri === Constants.URL_QUERY_INFO_SCHEME);
+        generateInitialQueryString(property, defaultInitialString, resultObject, mpdUrlQuery);
+        resultObject.finalQueryString = generateFinalQueryString(resultObject.initialQueryString, property);
+        resultObject.sameOriginOnly = property?.ExtUrlQueryInfo?.sameOriginOnly;
+        resultObject.includeInRequests = property?.ExtUrlQueryInfo?.includeInRequests ? property?.ExtUrlQueryInfo?.includeInRequests?.split(' ') : ['segment'];
+        resultObject.queryParams = parseQueryParams(resultObject?.initialQueryString);
+    }
+
+    function createFinalQueryStrings(manifest) {
+        mpd = {};
+        const manifestUrl = new URL(manifest.url);
+        mpd.origin = manifestUrl.origin;
+        mpd.period = [];
+        mpd.finalQueryString = '';
+        const mpdUrlQuery = manifest.url.split('?')[1];
+
+        generateQueryParams(mpd, manifest, 'SupplementalProperty', mpdUrlQuery, '');
```

Comment:
I don't quite understand this function. The `EssentialProperty` descriptor can also be present at MPD level from my understanding. In addition, the `SupplementalProperty` descriptor can be present on `AdaptationSet` and `Representation` level.

From the 23009-1:

> As defined by this specification, each of these descriptors may be present at the MPD, Adaptation Set or at the Representation level. Only SupplementalProperty descriptor may be present at the Period level.